### PR TITLE
[JUJU-3306] Ensure that a target controller has the expected external controller records created

### PR DESCRIPTION
**Pull request by wallyworld**
If a source controller hosts consuming and offering models with a cross model relation, then migrating the consuming model to a new controller breaks because the model on the target controller now has relation(s) to an external controller (it previously was the same controller). But the migration process does not create the necessary external controller record.

Here we extend what the final model activate api call does - it is given the details of the source controller and any hosted models on that controller to which the migrated model has relations. This info is used to create the external controller record. We also need to ensure that any relevant remote application records has the source controller correctly set for reference counting purposes.

The migration target facade gets a version bump. We need to add a precheck so that if the target controller does not have the version needed, and there is an external controller record that would need to be created, the migration fails.

Also removed unused code.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

bootstrap
add-model o
deploy juju-qa-dummy-source
offer dummy-source:sink
add-model c
deploy juju-qa-dummy-sink
relate dummy-sink o.dummy-source

bootstrap a new controller
migrate model c to new controller
check status

check logs on target controller to ensure no errors - previously the remote application worker would error
check the db to see that an external controller record is created and that the remote app has the source controller uuid set

## Bug reference

https://bugs.launchpad.net/bugs/2011662

### CODE REVIEW

File: rpc/params/migration.go [hunk 1]
```diff
@@ -121,6 +131,18 @@ type ModelArgs struct {
...
+// ActivateModelArgs holds args used to
+// activate a newly migrated model.
+type ActivateModelArgs struct {
+	ModelTag string `json:"model-tag"`
+
+	ControllerTag   string   `json:"controller-tag"`
+	ControllerAlias string   `json:"controller-alias,omitempty"`
+	SourceAPIAddrs  []string `json:"source-api-addrs"`
+	SourceCACert    string   `json:"source-ca-cert"`
+	CrossModelUUIDs []string `json:"cross-model-uuids"`
```
Comment (manadart) [inline rpc/params/migration.go:143]: Love comments for these public members too.

File: apiserver/facades/controller/migrationtarget/migrationtarget.go [hunk 2]
```diff
@@ -162,13 +168,71 @@ func (api *API) Abort(args params.ModelArgs) error {
...
+
+	if err := model.SetStatus(status.StatusInfo{Status: status.Available}); err != nil {
+		return errors.Trace(err)
+	}
+
+	// TODO(fwereade) - need to validate binaries here.
+	return model.SetMigrationMode(state.MigrationModeNone)
+}
+
+// Activate sets the migration mode of the model to "none", meaning it
```
Comment (manadart) [inline apiserver/facades/controller/migrationtarget/migrationtarget.go:186]: Does more than this now.

**Review by manadart — COMMENTED** [review]
Code looks good. Nicely done there.

Regarding QA, I observed these errors in the log during migration:
```
controller-0: 16:34:05 INFO juju.worker.machineundertaker tearing down machine undertaker
controller-0: 16:34:05 ERROR juju.worker.remoterelations cmr error in relation units worker for dummy-source:sink dummy-sink:source: connection is shut down
controller-0: 16:34:05 INFO juju.worker.logger logger worker stopped
controller-0: 16:34:05 ERROR juju.worker.remoterelations cmr error in remote relations worker for relation dummy-source:sink dummy-sink:source: connection is shut down
controller-0: 16:34:05 ERROR juju.worker.remoterelations cmr error in relation units worker for dummy-source:sink dummy-sink:source: connection is shut down
controller-0: 16:34:05 ERROR juju.worker.remoterelations cmr error in remote relations worker for relation dummy-source:sink dummy-sink:source: connection is shut down
controller-0: 16:34:05 INFO juju.worker.provisioner exiting provisioner task loop; err: catacomb 0xc002d06560 is dying
unit-dummy-sink-0: 16:34:05 INFO juju.worker.migrationminion migration phase is now: QUIESCE
machine-0: 16:34:05 INFO juju.worker.migrationminion migration phase is now: QUIESCE
controller-0: 16:34:05 ERROR juju.worker.remoterelations cmr error in remote application worker for dummy-source: connection is shut down
```

The offer ended up in a blocked state.
```
Model  Controller  Cloud/Region  Version   SLA          Timestamp
c      dst         lxd/default   2.9.43.1  unsupported  16:38:57+01:00

SAAS          Status   Store  URL
dummy-source  blocked  lxd    admin/o.dummy-source

App         Version  Status   Scale  Charm               Channel        Rev  Exposed  Message
dummy-sink           waiting      1  juju-qa-dummy-sink  latest/stable    7  no       Waiting for token

Unit           Workload  Agent  Machine  Public address  Ports  Message
dummy-sink/0*  waiting   idle   0        10.246.27.171          Waiting for token

Machine  State    Address        Inst id        Series  AZ  Message
0        started  10.246.27.171  juju-bdc764-0  xenial      Running
```

File: rpc/params/migration.go (hunk 1 above)
Comment (wallyworld) [reply rpc/params/migration.go:143]: added, was just following convention for other structs in this file.
i do think they are self explanatory though. most of the rpc params struct members are uncommented in general, and the comments which are there are extremely pithy

**Pull request merged**